proto/examples/sasltest/main.rs

```diff
@@ -0,0 +1,167 @@
+
+use ldap3_proto::proto::*;
+use ldap3_proto::LdapCodec;
+
+#[tokio::main]
+async fn main() -> Result<(), ()> {
+    let subs = tracing_subscriber::FmtSubscriber::builder()
+        .with_max_level(Level::DEBUG)
+        .finish();
+    tracing::subscriber::set_global_default(subs).expect("setting default subscriber failed");
+
+    let ldap_password = env::var("LDAP_PASSWORD").unwrap(); // password
+    let ldap_server_addr = env::var("LDAP_SERVER_ADDR").unwrap(); // domain.com:port
+    let ldap_username = env::var("LDAP_USERNAME").unwrap(); // username@domain
+    let addr = SocketAddr::from_str(&ldap_server_addr).expect(&format!(
+        "Unable to parse address, addr is {:?}",
+        &ldap_server_addr
+    ));
+
+    let tcpstream = TcpStream::connect(addr)
+        .await
+        .map_err(|e| eprintln!("Failed to connect to -> {:?}", e))?;
+
+    let mut framed = Framed::new(tcpstream, LdapCodec::default());
+
+    let mut ntlm = AuthProvier::new(&ldap_username, &ldap_password);
+    let ntlm_token = ntlm.step(&[]).unwrap();
+
+    let msg = LdapMsg {
+        msgid: 1,
+        op: LdapOp::BindRequest(LdapBindRequest {
+            dn: "".to_string(),
+            cred: LdapBindCred::SASL(SaslCredentials {
+                mechanism: "GSS-SPNEGO".to_string(),
+                credentials: ntlm_token,
+            }),
+        }),
+        ctrl: vec![],
+    };
+
+    let _ = framed.send(msg).await.map_err(|e| {
+        eprintln!("Unable to send bind -> {:?}", e);
+    })?;
+    loop {
+        if let Some(Ok(msg)) = dbg!(framed.next().await) {
+            if let LdapOp::BindResponse(res) = dbg!(msg.op) {
+                if res.res.code == LdapResultCode::Success {
+                    break;
+                } else if res.res.code == LdapResultCode::SaslBindInProgress {
+                    if let Some(ref cred) = res.saslcreds {
+                        let ntlm_token = ntlm.step(cred).unwrap();
+                        let msg = LdapMsg {
+                            msgid: 2,
+                            op: LdapOp::BindRequest(LdapBindRequest {
+                                dn: "".to_string(),
+                                cred: LdapBindCred::SASL(SaslCredentials {
+                                    mechanism: "GSS-SPNEGO".to_string(),
+                                    credentials: ntlm_token,
+                                }),
+                            }),
+                            ctrl: vec![],
+                        };
+
+                        let _ = framed.send(msg).await.map_err(|e| {
+                            eprintln!("Unable to send bind -> {:?}", e);
+                        })?;
+                    }
+                    dbg!(res);
+                } else {
+                    panic!("Bind failed: {:?}", res)
+                }
+            }
+        } else {
+            panic!("Unable to get bind response")
+        }
+    }
+
+    print!("Bind successful");

```

Comment:
Nitpick: `println!` seems more appropriate in this case. Terminating a program without the newline may not be rendered in a very elegant manner depending on the environment.